reco_utils/common/tf_utils.py

```diff
@@ -0,0 +1,408 @@
+        wide_columns = _build_wide_columns(user_ids, item_ids)
+        deep_columns = _build_deep_columns(user_ids, item_ids, dnn_user_dim, dnn_item_dim, item_feat_col, item_feat_shape)
+        model = tf.estimator.DNNLinearCombinedRegressor(
+            model_dir=model_dir,
+            config=config,
+            # wide settings
+            linear_feature_columns=wide_columns,
+            linear_optimizer=_build_optimizer(linear_optimizer, linear_optimizer_lr, linear_l1_reg),
+            # deep settings
+            dnn_feature_columns=deep_columns,
+            dnn_hidden_units=dnn_hidden_units,
+            dnn_optimizer=_build_optimizer(dnn_optimizer, dnn_optimizer_lr),
+            dnn_dropout=dnn_dropout,
+            batch_norm=dnn_batch_norm
+        )
+    else:
+        raise ValueError("Model type should be either 'wide', 'deep', or 'wide_deep'")
+
+    return model, wide_columns, deep_columns
+
+
+def evaluation_log_hook(
+    estimator,
+    logger,
+    true_df,
+    y_col,
+    eval_df,
+    every_n_iter=10000,
+    model_dir=None,
+    batch_size=256,
+    eval_fns=None,
+    **eval_kwargs
+):
+    """Evaluation log hook for TensorFlow high-levmodel_direl API Estimator.
+    Note, to evaluate the model in the middle of training (by using this hook),
+    the model checkpointing steps should be equal or larger than the hook's since
+    TensorFlow Estimator uses the last checkpoint for evaluation or prediction.
+    Checkpoint frequency can be set via Estimator's run config.
+
+    Args:
+        estimator (tf.estimator.Estimator): Model to evaluate.
+        logger (Logger): Custom logger to log the results. E.g., define a subclass of Logger for AzureML logging.
+        true_df (pd.DataFrame): Ground-truth data.
+        y_col (str): Label column name in true_df
+        eval_df (pd.DataFrame): Evaluation data. May not include the label column as
+            some evaluation functions do not allow.
+        every_n_iter (int): Evaluation frequency (steps). Should be equal or larger than checkpointing steps.
+        model_dir (str): Model directory to save the summaries to. If None, does not record.
+        batch_size (int): Number of samples fed into the model at a time.
+            Note, the batch size doesn't affect on evaluation results.
+        eval_fns (iterable of functions): List of evaluation functions that have signature of
+            (true_df, prediction_df, **eval_kwargs)->(float). If None, loss is calculated on true_df.
+        **eval_kwargs: Evaluation function's keyword arguments. Note, prediction column name should be 'prediction'
+
+    Returns:
+        (tf.train.SessionRunHook) Session run hook to evaluate the model while training.
+    """
+
+    return _TrainLogHook(
+        estimator,
+        logger,
+        true_df,
+        y_col,
+        eval_df,
+        every_n_iter,
+        model_dir,
+        batch_size,
+        eval_fns,
+        **eval_kwargs
+    )
+
+
+class Logger(abc.ABC):
```

Comment:
where is this used?

Reply:
This is for training log hook to record our metrics while training by using any loggers.
Example of loggers we may want to use are AzureML's Run.log, papermill.record, etc.
(main use-case for now is AzureML's Run.log, but didn't want to make the log hook dependent to AzureML)